rfcs/0000-safe-transmute.md

```diff
@@ -0,0 +1,2160 @@
+A struct or enum variant is *fully implicitly constructable* at a given location only if, at that location, that type can be instantiated via its *implicit constructor*, and its fields are also *implicitly constructable*.
+
+The *implicit constructor* of a struct or enum variant is the constructor Rust creates implicitly from its definition; e.g.:
+```rust
+struct Point<T> {
+    x: T,
+    y: T,
+}
+
+let p = Point { x: 4, y: 2 };
+     // ^^^^^^^^^^^^^^^^^^^^ An instance of `Point` is created here, via its implicit constructor.
+```
+
+Limiting implicit constructability is the fundamental mechanism with which type authors build safe abstractions for `unsafe` code, whose soundness is dependent on preserving invariants on fields. Usually, this takes the form of restricting the visibility of fields. For instance, consider the type `NonEmptySlice`, which enforces a validity constraint on its fields via its constructor:
+
+```rust
+pub mod crate_a {
+
+    #[repr(C)]
+    pub struct NonEmptySlice<T> {
+        data: *const T,
+        len: usize,
+    }
+
+    impl<T> NonEmptySlice<T> {
+        pub fn from_array<const N: usize>(arr: &[T; N], len: usize) -> Self {
+            assert!(len <= N);
+            assert!(len > 0);
+            Self {
+                data: arr as *const T,
+                len,
+            }
+        }
+
+        pub fn first(&self) -> &T {
+            unsafe { &*self.data }
+        }
+    }
+
+}
+```
+It is sound for `first` to be a *safe* method is because the `from_array` constructor ensures that `data` is safe to dereference, and because `from_array` is the *only* way to safely initialize `NonEmptySlice` outside of `crate_a` (note that `NonEmptySlice`'s fields are *not* `pub`). As a rule: any field that is not marked `pub` should be assumed to be private *because* it is subject to safety invariants.
+
+Unfortunately, field visibility modifiers are not a surefire indicator of whether a type is *fully* implicitly constructable. A type author may restrict the implicit constructability of a type even in situations where all fields of that type (*and all fields of those fields*) are `pub`; consider:
+```rust
+pub mod crate_a {
+
+    #[repr(C)]
+    pub struct NonEmptySlice<T>(pub private::NonEmptySliceInner<T>);
+
+    impl<T> NonEmptySlice<T> {
+        pub fn from_array<const N: usize>(arr: &[T; N], len: usize) -> Self {
+            assert!(len <= N && len > 0);
+            Self(
+                private::NonEmptySliceInner {
+                    data: arr as *const T,
+                    len,
+                }
+            )
+        }
+
+        pub fn first(&self) -> &T {
+            unsafe { &*self.0.data }
+        }
+    }
+
+    // introduce a private module to avoid `private_in_public` error (E0446):
+    pub(crate) mod private {
+        #[repr(C)]
+        pub struct NonEmptySliceInner<T> {
+            pub data: *const T,
+            pub len: usize,
+        }
+    }
+
+}
```

Comment:
```suggestion
pub mod crate_a {

    #[repr(C)]
    pub struct NonEmptySlice<'a, T>(pub private::NonEmptySliceInner<'a, T>);

    impl<'a, T> NonEmptySlice<'a, T> {
        pub fn from_array<const N: usize>(arr: &'a [T; N], len: usize) -> Self {
            assert!(len <= N && len > 0);
            Self(
                private::NonEmptySliceInner {
                    data: arr as *const T,
                    len,
                    lifetime: core::marker::PhantomData,
                }
            )
        }

        pub fn first(&self) -> &'a T {
            unsafe { &*self.0.data }
        }
    }

    // introduce a private module to avoid `private_in_public` error (E0446):
    pub(crate) mod private {
        #[repr(C)]
        pub struct NonEmptySliceInner<'a, T> {
            pub data: *const T,
            pub len: usize,
            pub lifetime: core::marker::PhantomData<&'a ()>,
        }
    }

}
```